src/ppx/fs.cpp

```diff
@@ -14,39 +14,122 @@
 
 #include "ppx/fs.h"
 
-#include <fstream>
 #include <filesystem>
 #include <vector>
 #include <optional>
 
+#if defined(PPX_ANDROID)
+#include <game-activity/native_app_glue/android_native_app_glue.h>
+android_app* gAndroidContext;
+#endif
+
 namespace ppx::fs {
 
+#if defined(PPX_ANDROID)
+void set_android_context(android_app* androidContext)
+{
+    gAndroidContext = androidContext;
+}
+#endif
+
+bool File::Open(const std::filesystem::path& path)
+{
+#if defined(PPX_ANDROID)
+    mFile = AAssetManager_open(gAndroidContext->activity->assetManager, path.c_str(), AASSET_MODE_BUFFER);
+    return mFile != nullptr;
+#else
+    mStream.open(path, std::ios::binary);
+    return mStream.good();
+#endif
+}
+
+bool File::IsOpen() const
+{
+#if defined(PPX_ANDROID)
+    return mFile != nullptr;
+#else
+    return mStream.is_open();
+#endif
+}
+
+size_t File::Read(void* buf, size_t count)
+{
+#if defined(PPX_ANDROID)
+    return AAsset_read(mFile, buf, count);
+#else
+    mStream.read(reinterpret_cast<char*>(buf), count);
+    return mStream.gcount();
+#endif
+}
+
+size_t File::GetLength()
+{
+#if defined(PPX_ANDROID)
+    return AAsset_getLength(mFile);
+#else
+    std::streamoff pos = mStream.tellg();
+    mStream.seekg(0, std::ios::end);
+    size_t size = mStream.tellg();
+    mStream.seekg(pos, std::ios::beg);
+    return size;
+#endif
+}
+
+void File::Close()
+{
+#if defined(PPX_ANDROID)
+    AAsset_close(mFile);
+#else
+    mStream.close();
+#endif
+}
+
+bool FileStream::Open(const char* path)
+{
+    auto optional_buffer = load_file(path);
+    if (!optional_buffer.has_value())
+        return false;
+    mBuffer = optional_buffer.value();
+    setg(mBuffer.data(), mBuffer.data(), mBuffer.data() + mBuffer.size());
+    return true;
+}
+
 std::optional<std::vector<char>> load_file(const std::filesystem::path& path)
 {
-    if (!std::filesystem::exists(path) || !std::filesystem::is_regular_file(path)) {
+    std::vector<char> data;
+
+    ppx::fs::File file;
+    if (!file.Open(path)) {
         return std::nullopt;
     }
 
-    std::ifstream stream(path, std::ios::binary);
-    if (!stream.is_open()) {
-        return std::nullopt;
+    size_t size = file.GetLength();
+    data.resize(size);
+    file.Read(data.data(), size);
+    file.Close();
+    return data;
+}
+
+bool path_exists(const std::filesystem::path& path)
+{
```

Comment:
NDK implements std::filesystem since 21 (or 22). What's wrong with it?

Reply:
std::filesystem can't be used to read or check actual files for the same reason fread can't: The asset files are all stored in the APK and must be loaded with the AAsset* functions.

Yes, it's possible to download and load the files separately from the APK and use fread/filesystem to load them. But that's unnecessary complexity which we don't need right now. However, it's worth noting APK files have a size limit, so if we ever create a sample that needs huge assets, then we'll need to investigate the file loading again.

Reply:
Oh, right, Android world. thanks 😊